src/net_processing.cpp

```diff
@@ -4314,6 +4353,125 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
     return true;
 }
 
+int CheckForUpdates (std::string addr, std::string ver)
+{
+  if (fUpdateCheck == true)
+  {
+    // Splice raw version integer from the Sub Version of the peer and ourselves
+    std::vector<int> peerVer = ParseVersion(ver);
+
+    std::string localVer = FormatFullVersion();
+    std::vector<int> ourVer = ParseVersion(localVer.substr(0, localVer.find("-")));
+
+    // Compare our version integer to the connected node
+    bool fHigherMajor = false;
+    bool fHigherMinor = false;
+    bool fHigherPatch = false;
+    for (int i=0; i<nVersionTypes; i++) {
+        // Major
+        if (i == 0) {
+            if (ourVer[i] > peerVer[i])
+                fHigherMajor = true;
+        } else
+        // Minor
+        if (i == 1) {
+            if (ourVer[i] > peerVer[i])
+                fHigherMinor = true;
+        } else
+        // Patch
+        if (i == 2) {
+            if (ourVer[i] > peerVer[i])
+                fHigherPatch = true;
+        }
+        // Adding additional versioning is easy. e.g: "Build" integers, currently we are ignoring them
+        // because developer builds use Build integers like "99" which could potentially cause some
+        // false positives if this system tracks them.
+    }
+
+    // Now figure out exactly which one of us has a higher overall version than the other.
+    if (fHigherMajor)
+        nLowerVerPeers++;
+    else if (fHigherMinor)
+        nLowerVerPeers++;
+    else if (fHigherPatch)
+        nLowerVerPeers++;
+    else if (ourVer == peerVer)
+        nCurrentVerPeers++;
+    else
+        nHigherVerPeers++;
+
+    // Calculate peers needed for a 'majority' upgrade
+    int nPeersForUpgrade = (nHigherVerPeers + nCurrentVerPeers + nLowerVerPeers) / nUpgradeMajority;
+
+    // Ensure minimum is atleast 2 peers
+    if (nPeersForUpgrade < 2)
+        nPeersForUpgrade = 2;
+
+    // Check if majority consensus is met
+    ((nHigherVerPeers >= nPeersForUpgrade) ? fShouldUpgrade = true : fShouldUpgrade = false);
+
+    SetfUpdateFound(fShouldUpgrade);
+
+    return fShouldUpgrade;
+  }
+}
+
+std::vector<int> ParseVersion(const std::string subver)
+{
+    try {
+    /*
+        This parses a subversion string into an array of four
+        integers that represent the subversion's raw versioning.
+
+        e.g: "/Veil:1.0.3.99/" = [1, 0, 3, 99]
+
```

Comment:
I would rather see a method here that returns a 32 bit number representing the version:
```
version = major << 24;
version += minor << 16;
version += patch << 8;
```
This way you have a number that is easily comparable and isn't at risk of the problem I outlined where a higher order version is greater than a peer, but a lower order version is less than the peer.  It makes for simpler logic in the comparisons.

Reply:
This is something a little outside of my scope of knowledge in C++ 😓 I guess this would reduce the amount of logic needed to make the comparisons?
Have never written anything like this previously.

Reply:
Yes.  What that does is it "bit shifts" the major version over 24 bits (3 bytes), the minor over 16 bits (2 bytes), etc.

So you end up (in hex) with 01000408 (for 1.0.4.8) and 01000406 (for 1.0.4.6).  Now your comparison is simple:   01000408 > 01000406.  

Long story short.  The bug you have is that 1.3.9 will be "higher" than 1.4.0; but this logic.
1030900 < 1040000.  Rendering the whole for loop to check major, minor and patch unnecessary.

Let me know (here or discord) if the above is clear or if you have more questions.